docs/ref/optimizers/ceres_tutorial.md

```diff
@@ -0,0 +1,205 @@
+# Ceres Tutorial
+As with all nonlinear-optimizers they can be difficult to master, in the following we give examples on how to effectively implement custom cost functions in Ceres-Solver. In particular we will cover several different topics:
+
+- How to define a cost function?
+- How to define the optimization problem?
+- How do I define a numerical cost function?
+- How do I define an analytical cost function?
+- How do I load the optimization data?
+
+
+## How to define a cost function?
+Ceres an solve bounds constrained robustified non-linear least squares problems of the form:
+
+\begin{equation}
+  \min_{x} \dfrac{1}{2} \sum_{i} \rho_{i} (|| f_{i} (x_{i}, \dots, x_{i_{k}}) ||^{2})
+\end{equation}
+
+The expression $\rho_{i} (|| f_{i} (x_{i}, \dots, x_{i_{k}}) ||^{2})$ is known in ceres as a **`ResidualBlock`**, where $f_{i}(\cdot)$ is a cost function that depends on the parameter blocks $[x_{i}, \dots, x_{i_{k}}]$.
+
+To get started, consider the problem of finding the minimum of the function
+
+\begin{equation}
+  \dfrac{1}{2} (10 - x)^{2}
+\end{equation}
+
+The cost function in this case is:
+
+\begin{equation}
+  f(x) = (10 - x)
+\end{equation}
+
+This is a trivial problem, whose minimum is located at $x = 10$. The first step
+is to write a functor that will evaluate this function:
+
+    struct CostFunctor {
+      template <typename T>
+      bool operator()(const T* const x, T* residual) const {
+        residual[0] = T(10.0) - x[0];
+        return true;
+      }
+    };
+
+Keypoints to the example code:
+
+- In Ceres a cost function has to be a functor (function objects: a plain old C++ object plus the `()` operator), this is essentially so that the ceres solver can call your cost function with the overridded `()` operator you implemented, this is where you will add your custom cost function, for more information on functors see [this][stack_overflow-functors].
+- The output of your cost function is written to `residual[0]`, in all cases `residual` is always an array of type `T`, further in this example we are only writing to `residual[0]` because the output dimension for our scalar cost function is 1.
```

Comment:
Try to be clear on the difference between cost function and residual throughout.  We only have to define the residual, because the cost function always is constructed the same way (cost = sum (residuals)^2).